### PR TITLE
consensus/istanbul: verify block with header signatures in synching

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: consensus/istanbul/backends/simple/engine.go [hunk 1]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+		if len(istanbul.CommittedSeal) == 0 {
+			return errInvalidCommittedSeals
+		}
+
+		// Checks whether the committed seals are generated by parent validators
+		validSeal := 0
+		// 1. Getting committed seals from current header
+		for _, seal := range istanbul.CommittedSeal {
+			// 2. gets the original address by using seal and parent block hash
+			addr, err := sb.getSignatureAddress(parent.Hash().Bytes(), seal)
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: current

File: consensus/istanbul/backends/simple/engine.go [hunk 2]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+				if parentValidator == addr {
+					sb.logger.Info("found a valid addr", "addr", addr.Hex())
+					validSeal += 1
+					break
+				}
+			}
+		}
+
+		// The length of validSeal should large than number of faulty node +1
+		if validSeal > int(sb.core.NumberOfFaultyNode())+1 {
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: validSeal > int(sb.core.NumberOfFaultyNode())

File: consensus/istanbul/backends/simple/engine.go [hunk 3]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+		for _, seal := range istanbul.CommittedSeal {
+			// 2. gets the original address by using seal and parent block hash
+			addr, err := sb.getSignatureAddress(parent.Hash().Bytes(), seal)
+			if err != nil {
+				sb.logger.Error("not a valid addr", "err", err)
+				continue
+			}
+
+			// 3. checks whether the address is appeared in parent validators
+			for _, parentValidator := range parentIstanbul.Validators {
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: replace parentIstanbul with parentExtra

File: consensus/istanbul/backends/simple/engine.go [hunk 4]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+		}
+
+		// Checks whether the committed seals are generated by parent validators
+		validSeal := 0
+		// 1. Getting committed seals from current header
+		for _, seal := range istanbul.CommittedSeal {
+			// 2. gets the original address by using seal and parent block hash
+			addr, err := sb.getSignatureAddress(parent.Hash().Bytes(), seal)
+			if err != nil {
+				sb.logger.Error("not a valid addr", "err", err)
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: not a valid addrress

File: consensus/istanbul/backends/simple/engine.go [hunk 5]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+				continue
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: return

File: consensus/istanbul/backends/simple/engine.go [hunk 6]
```diff
@@ -150,6 +152,40 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
...
+			}
+
+			// 3. checks whether the address is appeared in parent validators
+			for _, parentValidator := range parentIstanbul.Validators {
+				if parentValidator == addr {
+					sb.logger.Info("found a valid addr", "addr", addr.Hex())
+					validSeal += 1
+					break
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/engine.go]: return

File: consensus/istanbul/core/core.go [hunk 7]
```diff
@@ -41,8 +41,8 @@ func New(backend istanbul.Backend, config *istanbul.Config) Engine {
 	return &core{
 		config:            config,
 		address:           backend.Address(),
-		N:                 n,
-		F:                 f,
+		n:                 n,
```
Comment (bailantaotao) [inline consensus/istanbul/core/core.go]: move N and F into validatorSet

File: consensus/istanbul/backends/simple/backend.go [hunk 8]
```diff
@@ -162,7 +162,13 @@ func (sb *simpleBackend) Verify(proposal istanbul.Proposal) error {
 		return errInvalidProposal
 	}
 	// verify the header of proposed block
-	return sb.VerifyHeader(sb.chain, block.Header(), false)
+	if err := sb.VerifyHeader(sb.chain, block.Header(), false); err != nil {
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/backend.go]:
```go
err := sb.VerifyHeader(sb.chain, block.Header(), false)
// some comments ...
if err != nil && err != errEmptyCommittedSeals {
        return err
}
return nil
```

I suggest reorganizing the `if` condition to make it more clear.

File: consensus/istanbul/backends/simple/engine.go [hunk 9]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
 	return nil
 }
 
+// verifyCommittedSeals checks whether the committed seal is signed by parent's validator
+func (sb *simpleBackend) verifyCommittedSeals(header, parent *types.Header) error {
+	extra := types.ExtractToIstanbul(header)
+	parentExtra := types.ExtractToIstanbul(parent)
+
+	// The length of Committed seals should lager than 0
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: should be larger than

File: consensus/istanbul/backends/simple/backend.go [hunk 10]
```diff
@@ -162,7 +162,13 @@ func (sb *simpleBackend) Verify(proposal istanbul.Proposal) error {
...
+		if err == errEmptyCommittedSeals {
```
Comment (tailingchen) [inline consensus/istanbul/backends/simple/backend.go]: Add comment to explain why we ignore `errEmptyCommittedSeals ` here.

File: consensus/istanbul/backends/simple/engine.go [hunk 11]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+	if len(extra.CommittedSeal) == 0 {
+		return errEmptyCommittedSeals
+	}
+
+	// Checks whether the committed seals are generated by parent validators
+	validSeal := 0
+	// 1. Getting committed seals from current header
+	for _, seal := range extra.CommittedSeal {
+		// 2. gets the original address by using seal and parent block hash
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: Get the original address by seal and parent block hash.

File: consensus/istanbul/backends/simple/engine.go [hunk 12]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+		proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
+		addr, err := sb.getSignatureAddress(proposalSeal, seal)
+		if err != nil {
+			sb.logger.Error("not a valid address", "err", err)
+			return errInvalidSignature
+		}
+
+		found := false
+		// 3. checks whether the address is appeared in parent validators
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: 3. Check whether the address is in parent validator set.

File: consensus/istanbul/backends/simple/engine.go [hunk 13]
```diff
@@ -63,6 +63,10 @@ var (
 	errInconsistentValidatorSet = errors.New("non empty uncle hash")
 	// errInvalidTimestamp is returned if the timestamp of a block is lower than the previous block's timestamp + the minimum block period.
 	errInvalidTimestamp = errors.New("invalid timestamp")
+	// errInvalidCommittedSeals is returned if the field of committed seals is not appeared in parent validators.
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: errInvalidCommittedSeals is returned if the committed seal is not signed by any of parent validators.

File: consensus/istanbul/backends/simple/engine.go [hunk 14]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+			}
+		}
+
+		// 4. returns error if seal is not signed by parent validator.
+		if !found {
+			return errInvalidCommittedSeals
+		}
+	}
+
+	// The length of validSeal should large than number of faulty node + 1
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: should be larger than

File: consensus/istanbul/backends/simple/engine.go [hunk 15]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+	parentExtra := types.ExtractToIstanbul(parent)
+
+	// The length of Committed seals should lager than 0
+	if len(extra.CommittedSeal) == 0 {
+		return errEmptyCommittedSeals
+	}
+
+	// Checks whether the committed seals are generated by parent validators
+	validSeal := 0
+	// 1. Getting committed seals from current header
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: Get committed seals

File: consensus/istanbul/validator.go [hunk 16]
```diff
@@ -41,4 +41,6 @@ type ValidatorSet interface {
 	GetProposer() Validator
 	// Check whether the validator with given address is a proposer
 	IsProposer(address common.Address) bool
+	// Get the maximum number of faulty node
```
Comment (alanchchen) [inline consensus/istanbul/validator.go]: faulty nodes

File: consensus/istanbul/backends/simple/engine.go [hunk 17]
```diff
@@ -209,6 +218,52 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
 	return nil
 }
 
+// verifyCommittedSeals checks whether the committed seal is signed by parent's validator
+func (sb *simpleBackend) verifyCommittedSeals(header, parent *types.Header) error {
```
Comment (alanchchen) [inline consensus/istanbul/backends/simple/engine.go]: We should check if committed seals are signed by F+1 **different** validators.

File: consensus/istanbul/core/core.go [hunk 18]
```diff
@@ -282,3 +282,8 @@ func prepareProposalSeal(hash common.Hash, msgCode uint64) []byte {
 	buf.Write([]byte{byte(msgCode)})
 	return buf.Bytes()
 }
+
+// PrepareCommittedSeal returns a slice from the given hash.
```
Comment (alanchchen) [inline consensus/istanbul/core/core.go]: Slice of what ?

File: consensus/istanbul/core/core.go (hunk 7 above)
Comment (markya0616) [inline consensus/istanbul/core/core.go]: OK

File: consensus/istanbul/backends/simple/backend.go (hunk 8 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/backend.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 9 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/backends/simple/backend.go (hunk 10 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/backend.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 11 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 12 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 13 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 14 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 15 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/validator.go (hunk 16 above)
Comment (markya0616) [inline consensus/istanbul/validator.go]: OK

File: consensus/istanbul/backends/simple/engine.go (hunk 17 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: OK

File: consensus/istanbul/core/core.go (hunk 18 above)
Comment (markya0616) [inline consensus/istanbul/core/core.go]: OK

File: consensus/istanbul/backend/engine.go [hunk 19]
```diff
@@ -236,6 +239,53 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
 
+// verifyCommittedSeals checks whether the committed seal is signed by parent's validator
+func (sb *simpleBackend) verifyCommittedSeals(chain consensus.ChainReader, header *types.Header, parents []*types.Header) error {
+	number := header.Number.Uint64()
+	// We don't need to verify committed seals in the genesis block
+	if number == 0 {
+		return nil
+	}
+
+	// Retrieve the snapshot needed to verify this header and cache it
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: What does this mean?

Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: 抄PoA的   應該是指說把snapshot抓出來 然後需要用這snapshot去驗證header 並且會把他cache住

File: consensus/istanbul/backend/engine.go [hunk 20]
```diff
@@ -236,6 +239,53 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
 	return nil
 }
 
+// verifyCommittedSeals checks whether the committed seal is signed by parent's validator
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: every committed seal is signed by one of the parent's validators

File: consensus/istanbul/backend/engine.go [hunk 21]
```diff
@@ -236,6 +239,53 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+	}
+
+	extra := types.ExtractToIstanbul(header)
+	// The length of Committed seals should be larger than 0
+	if len(extra.CommittedSeal) == 0 {
+		return errEmptyCommittedSeals
+	}
+
+	validators := snap.ValSet.Copy()
+	// Checks whether the committed seals are generated by parent validators
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: // Check whether the committed seals are generated by parent's validators

Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 20 above)
Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go [hunk 22]
```diff
@@ -236,6 +239,53 @@ func (sb *simpleBackend) verifySigner(chain consensus.ChainReader, header *types
...
+	validSeal := 0
+	proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
+	// 1. Get committed seals from current header
+	for _, seal := range extra.CommittedSeal {
+		// 2. Get the original address by seal and parent block hash
+		addr, err := istanbul.GetSignatureAddress(proposalSeal, seal)
+		if err != nil {
+			sb.logger.Error("not a valid address", "err", err)
+			return errInvalidSignature
+		}
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: // Every validator can have only one seal. There are more than one seals signed by the same validator if `addr` cannot be found in `validators`.

Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: OK

**Pull request merged**